### PR TITLE
Add generic range cache (FE-1231)

### PR DESCRIPTION
**Pull request by hbenl**
and use it to reimplement the hitpoints cache.

The generic range cache can be used to cache anything that can be fetched for a point range. If parts of a requested range have already been cached, it will only fetch the missing parts of the requested range. It is the first step towards the new analysis cache.

### CODE REVIEW

**Review by bvaughn — COMMENTED** [review]
Do you mind holding off on landing this for a day or two? It touches a few of the points files– and while the conflicts would be minimal, I'd love to avoid making that PR any more complicated to land.

**Comment by hbenl** [comment]
@bvaughn No problem, I'm not in a hurry landing this.

**Comment by hbenl** [comment]
I still have to figure out how to teach the new `ReplayClient.streamAnalysis()` to the recorder/player for our replay-next tests.

File: packages/protocol/analysisManager.ts [hunk 1]
```diff
@@ -166,7 +167,10 @@ class AnalysisManager {
       this.handlers.set(analysisId, handler);
       await Promise.all([
         !handler.onAnalysisResult || client.Analysis.runAnalysis({ analysisId }, sessionId),
-        !handler.onAnalysisPoints || client.Analysis.findAnalysisPoints({ analysisId }, sessionId),
+        !(handler.onAnalysisPoints || handler.onPointsFinished) ||
+          client.Analysis.findAnalysisPoints({ analysisId }, sessionId).then(
+            handler.onPointsFinished
+          ),
```
Comment (bvaughn) [inline packages/protocol/analysisManager.ts:173]: This looks wrong, I think.

If you pass both `onAnalysisPoints` but not `onPointsFinished` – this will add a handler for `onPointsFinished`?

File: packages/shared/client/types.ts [hunk 2]
```diff
@@ -215,6 +211,14 @@ export interface ReplayClientInterface {
...
     onMatches: (matches: SearchSourceContentsMatch[], didOverflow: boolean) => void
   ): Promise<void>;
+  streamAnalysis(
+    params: AnalysisParams,
+    handlers: {
+      onPoints?: (points: PointDescription[]) => void;
+      onResults?: (results: AnalysisEntry[]) => void;
+      onError?: (error: any) => void;
+    }
+  ): { pointsReceived: Promise<void>; resultsReceived: Promise<void> };
```
Comment (bvaughn) [inline packages/shared/client/types.ts:221]: nit: Would "finished" be clearer than "received"? Seems like "received" could potentially mean either "first result received" or "all results received"

File: src/ui/components/FocusContextReduxAdapter.tsx [hunk 3]
```diff
@@ -24,6 +34,14 @@ export default function FocusContextReduxAdapter({ children }: PropsWithChildren
     });
   }, [focusRegion, loadedRegions]);
 
+  const rangeForAnalysis = useMemo(
+    () =>
+      deferredFocusRegion
+        ? { begin: deferredFocusRegion.begin.point, end: deferredFocusRegion.end.point }
+        : { begin: "0", end: endpoint },
+    [deferredFocusRegion, endpoint]
+  );
```
Comment (bvaughn) [inline src/ui/components/FocusContextReduxAdapter.tsx:43]: This just a convenience value? I guess that makes sense, since we're using it so many places.

**Review by bvaughn — COMMENTED** [review]
Haven't dug into the actual range code yet– is this ready for revieW?

File: packages/protocol/analysisManager.ts (hunk 1 above)
Comment (hbenl) [reply packages/protocol/analysisManager.ts:173]: The `onAnalysisPoints()` handler will be called by the `AnalysisManager.onAnalysisPoints()` method which receives the `Analysis.analysisPoints` events. But the backend will only send these events to us if we call the `Analysis.findAnalysisPoints` command, that's why the `onAnalysisPoints()` handler appears in this condition.

File: src/ui/components/FocusContextReduxAdapter.tsx (hunk 3 above)
Comment (hbenl) [reply src/ui/components/FocusContextReduxAdapter.tsx:43]: > This just a convenience value? I guess that makes sense, since we're using it so many places.

Exactly. I had put it in a hook at first but then I thought that it makes sense to add it to the `FocusContext` :shrug:

**Comment by hbenl** [comment]
> is this ready for revieW?

Yes, I just need to fix the tests before merging.

File: packages/protocol/analysisManager.ts (hunk 1 above)
Comment (bvaughn) [reply packages/protocol/analysisManager.ts:173]: I don't think I understand. Those things seem kind of far apart from each other to be an uncommented dependency in this function. (Also maybe it's just me but I really have trouble parsing the way this code is written.)

Could we write it without as much `!` and `||` to maybe make it more clearer?
Is this the same code?

```js
const promises: Promise<any>[] = [];
if (handler.onAnalysisResult) {
  promises.push(client.Analysis.runAnalysis({ analysisId }, sessionId));
}
if (handler.onAnalysisPoints && handler.onPointsFinished) {
  promises.push(
    client.Analysis.findAnalysisPoints({ analysisId }, sessionId).then(
      handler.onPointsFinished
    )
  );
}
await Promise.all(promises);
```

Comment (hbenl) [reply packages/protocol/analysisManager.ts:173]: Yes, this is the same code except we still need to say `if (handler.onAnalysisPoints || handler.onPointsFinished)` instead of `if (handler.onPointsFinished)`.

Comment (bvaughn) [reply packages/protocol/analysisManager.ts:173]: Can you replace the above with something like this then, for readability?
```js
const promises: Promise<any>[] = [];
if (handler.onAnalysisResult) {
  promises.push(client.Analysis.runAnalysis({ analysisId }, sessionId));
}
if (handler.onAnalysisPoints || handler.onPointsFinished) {
  // We check onAnalysisPoints too even though it isn't directly referenced here
  // because unless we call Analysis.findAnalysisPoints,
  // that callback won't be called.
  promises.push(
    client.Analysis.findAnalysisPoints({ analysisId }, sessionId).then(
      handler.onPointsFinished
    )
  );
}
await Promise.all(promises);
```

File: packages/replay-next/src/suspense/HitPointsCache.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+function createHitPointsCache(location: Location, condition: string | null) {
+  return createGenericRangeCache<TimeStampedPoint>(
+    async (client, range, cacheValues, cacheError) => {
+      const locations = client.getCorrespondingLocations(location).map(location => ({
+        location,
+      }));
+
+      if (condition) {
+        const mapper = `
```
Comment (bvaughn) [inline packages/replay-next/src/suspense/HitPointsCache.ts:33]: Shouldn't this be typed code now, after @markerikson's recent change?

File: packages/replay-next/src/utils/range-list.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+export function isInRange(b: bigint, range: Range) {
+  return b >= range.begin && b < range.end;
+}
+
+export function rangesOverlap(range1: Range, range2: Range) {
+  return range1.end > range2.begin && range1.begin < range2.end;
+}
+
+export function rangeContains(range1: Range, range2: Range) {
```
Comment (bvaughn) [inline packages/replay-next/src/utils/range-list.ts:22]: nit: Name one of these something like `containerRange` or something so it's clear which directly the args are expected in? (I often finds "contains" type functions confusing when I'm using them).

File: packages/replay-next/src/utils/range-list.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  return range1.begin <= range2.begin && range1.end >= range2.end;
+}
+
+export function getMissingRanges(requested: Range, cached: Range[]): Range[] {
```
Comment (bvaughn) [inline packages/replay-next/src/utils/range-list.ts:26]: Functions like this should have associated unit tests– particularly to cover edge cases.

File: packages/replay-next/src/utils/range-list.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+      begin: point,
+      end: min(cached[index].begin, requested.end),
+    });
+    point = cached[index].end;
+    index++;
+  }
+  return missing;
+}
+
+export function mergeRanges(ranges1: Range[], ranges2: Range[]): Range[] {
```
Comment (bvaughn) [inline packages/replay-next/src/utils/range-list.ts:54]: Same RE: unit tests

File: packages/replay-next/src/suspense/createGenericRangeCache.ts [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        requestedRange,
+        mergeRanges(store.cachedRanges, store.pendingRanges)
+      );
+
+      const failedRange = failedRanges.find(failedRange =>
+        missingRanges.some(missingRange => rangeContains(missingRange, failedRange.range))
+      );
+      if (failedRange) {
+        throw failedRange.error;
+      }
```
Comment (bvaughn) [inline packages/replay-next/src/suspense/createGenericRangeCache.ts:83]: Won't this potentially give us false positives?

If we try to fetch all events for a whole recording, and it fails, then wouldn't any subsequent focus ranges be flagged as failed?

File: packages/replay-next/src/suspense/createGenericRangeCache.ts [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+        pendingRange.promise = new Promise<void>(async (resolve, reject) => {
+          pendingRanges.push(pendingRange);
+          store.pendingRanges = mergeRanges(store.pendingRanges, [missingRange]);
+
+          let failed = false;
+          function onError(error: any) {
+            if (!failed) {
+              failed = true;
+              failedRanges.push({ range: missingRange, error });
```
Comment (bvaughn) [inline packages/replay-next/src/suspense/createGenericRangeCache.ts:99]: Maybe I'm reading this wrong, but I _think_ we should only gate the call to `reject` in `if (!failed)` but we should always push failed ranges?

These generic caches should have unit tests too I think, that explicitly cover edge cases like this.

File: packages/shared/client/ReplayClient.ts [hunk 10]
```diff
@@ -928,6 +776,32 @@ export class ReplayClient implements ReplayClientInterface {
...
+  ): { pointsReceived: Promise<void>; resultsReceived: Promise<void> } {
+    let resolvePointsReceived!: () => void;
+    const pointsReceived = new Promise<void>(resolve => (resolvePointsReceived = resolve));
+    let resolveResultsReceived!: () => void;
+    const resultsReceived = new Promise<void>(resolve => (resolveResultsReceived = resolve));
+
+    this._waitForRangeToBeLoaded(params.range || null).then(() =>
+      analysisManager.runAnalysis(params, {
+        onAnalysisPoints: handlers.onPoints,
+        onAnalysisResult: handlers.onResults,
```
Comment (bvaughn) [inline packages/shared/client/ReplayClient.ts:795]: Naming nit: Why is a param called "on results" (plural) re-mapped to something called "on result" (singular)?

File: packages/replay-next/src/suspense/HitPointsCache.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+        await client.streamAnalysis(
+          {
+            effectful: false,
+            locations,
+            mapper,
+            range,
+          },
+          {
+            onResults: results => {
```
Comment (bvaughn) [inline packages/replay-next/src/suspense/HitPointsCache.ts:77]: I know the analysis manager pre-dates this PR, but I'm confused about the difference between `onResult` (used here) and `onPoints` (used below). Looks like we could almost avoid this fork entirely (and just pass a mapper or an empty string) _except_ for that.

I guess one calls `Analysis.findAnalysisPoints` and the other calls `Analysis.runAnalysis`. Is one more efficient than the other?

File: packages/replay-next/src/suspense/createGenericRangeCache.ts (hunk 8 above)
Comment (hbenl) [reply packages/replay-next/src/suspense/createGenericRangeCache.ts:83]: No, it's the other way around: only ranges that completely contain a failed range will be flagged as failed. (Yeah, I find contains type functions confusing as well)

File: packages/replay-next/src/suspense/createGenericRangeCache.ts (hunk 9 above)
Comment (hbenl) [reply packages/replay-next/src/suspense/createGenericRangeCache.ts:99]: If `failed` is `true` then we have already pushed the `missingRange` to `failedRanges`, no need to do that again.

Comment (bvaughn) [reply packages/replay-next/src/suspense/createGenericRangeCache.ts:99]: That's only true for a single range though, right? Can't there be more than one?

File: packages/replay-next/src/suspense/HitPointsCache.ts (hunk 11 above)
Comment (hbenl) [reply packages/replay-next/src/suspense/HitPointsCache.ts:77]: Yes, finding analysis points is a lot faster than getting the results - the first is just a lookup in the backend while the second needs to run a browser to each point to execute the mapper. So `pointsFinished` will resolve a lot earlier than `resultsFinished`.
When we don't have a `condition` we only need to ask for the points but when we have one we need to execute it for every point to find out if that point should be included.

File: packages/replay-next/src/suspense/HitPointsCache.ts (hunk 4 above)
Comment (hbenl) [reply packages/replay-next/src/suspense/HitPointsCache.ts:37]: This mapper hadn't been converted yet.

File: packages/replay-next/src/suspense/createGenericRangeCache.ts (hunk 9 above)
Comment (hbenl) [reply packages/replay-next/src/suspense/createGenericRangeCache.ts:99]: Yes, there can be more than one, but since `failed` is declared within the loop for the ranges, each range gets its own `failed` flag.

Comment (bvaughn) [reply packages/replay-next/src/suspense/createGenericRangeCache.ts:99]: I see.

File: packages/replay-next/src/suspense/HitPointsCache.ts (hunk 11 above)
Comment (markerikson) [reply packages/replay-next/src/suspense/HitPointsCache.ts:81]: Yep, this is the underlying problem with hit points and print statements to begin with, and this is also why the "breakpoints" implementation that Josh and I had come up with mid-last-year intentionally separated out the "find points" and "run actual analysis" steps in that `protocol/thread/analysis.ts` file.

**Comment by bvaughn** [comment]
I forget where we landed on unit tests– but this PR adds a lot of new functionality and no new tests. (It actually _removes_ a pre-existing test.) Can you do a follow up PR with unit tests for this stuff, @hbenl?

**Comment by hbenl** [comment]
Yep, I created FE-1266 for that.
